experimental/storage/src/upload/types.ts

```diff
@@ -0,0 +1,68 @@
+import { PutObjectCommandInput, S3, S3Client, Tag as Tag_ } from "@aws-sdk/client-s3";
+
+export namespace Upload {
+  export interface Progress {
+    loaded?: number;
+    total?: number;
+    part?: number;
+    Key?: string;
+    Bucket?: string;
+  }
+
+  export interface S3Location {
+    Key: string;
+    Bucket: string;
+  }
+
+  export interface S3Target extends S3Location {
+    service: S3;
+  }
+  export interface S3ClientTarget extends S3Location {
+    service: S3Client;
+  }
+
+  export type Target = S3Target | S3ClientTarget;
+
+  export type Tag = Tag_;
+
+  export type SendData = PutObjectCommandInput["Body"];
+
+  export interface Configuration {
+    /**
+     * The size of the concurrent queue manager to upload parts in parallel. Set to 1 for synchronous uploading of parts. Note that the uploader will buffer at most queueSize * partSize bytes into memory at any given time.
+     * default: 4
+     */
+    queueSize: number;
+
+    /**
+     * Default: 5 mb
+     * The size in bytes for each individual part to be uploaded. Adjust the part size to ensure the number of parts does not exceed maxTotalParts. See minPartSize for the minimum allowed part size.
+     */
+    partSize: number;
+
+    /**
+     * Default: false
+     * Whether to abort the multipart upload if an error occurs. Set to true if you want to handle failures manually.
+     */
+    leavePartsOnError: boolean;
+
+    /**
+     * The tags to apply to the object.
+     */
+    tags: Tag_[];
+  }
+
+  // body : string | Buffer | Uint8Array | Blob | Readable
+  export interface Options extends Partial<Configuration> {
+    /**
+     * The "Body" parameter of a put request. This is the data that is uploaded.
+     */
+    data: SendData;
+
+    /**
+     * A list of valid target service clients.
+     * This the target where we upload data.
+     */
+    targets: Target[];
```

Comment:
Does it mean users can upload one stream to multiple object? I don't think we support it in v2. Is there a feature request for it? Conceptually, a managed upload is just uploading one big object in chunks. Of course we can support abstractions to upload multiple objects(folder) on top of it. But it's beyond the minimal requirement of a managed uploader.

Reply:
Yes exactly! You can upload the same data to many targets in parallel. For example if I wanted to archive and put to a different s3 bucket thats totally doable. I made this choice to make the API extensible.

Reply:
Why can't users just [tee](https://streams.spec.whatwg.org/#tee-a-readable-stream) the stream and call the upload on each of them? It will be easier for error handling. Also, as a user used to v2 multipart upload, changing target to an array is a little confusing. But if you feel this is useful, maybe type to `targes: Target | Target[]`

Reply:
I think this is a really fundamental point. Lets discuss in our meeting today. Like we've only allowing one target, we might as well call this `client: ` and then wrap the target data, like `key` and `bucket` with the data object.

I think it the API is better looking if we keep it as is, but I totally see your point. Like why not just have them call this twice?